### PR TITLE
refactor: tx multi payment pallet

### PR DESCRIPTION
**Pull request by enthusiastmartin**

### CODE REVIEW

File: pallets/transaction-multi-payment/src/lib.rs [hunk 1]
```diff
@@ -172,34 +131,29 @@ pub mod pallet {
...
-		/// Fallback price was not found.
-		FallbackPriceNotFound,
-
 		/// Math overflow
 		Overflow,
 	}
 
+	#[pallet::type_value]
+	pub fn DefaultFeeCurrency<T: Config>() -> T::AssetId {
+		T::NativeAssetId::get()
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:140]: default currency is different for EVM accounts, they have WETH

File: pallets/transaction-multi-payment/src/lib.rs [hunk 2]
```diff
@@ -245,12 +196,7 @@ pub mod pallet {
 		pub fn set_currency(origin: OriginFor<T>, currency: AssetIdOf<T>) -> DispatchResult {
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:196]: EVM account cannot change payment currency from default

File: pallets/transaction-multi-payment/src/lib.rs [hunk 3]
```diff
@@ -172,34 +131,29 @@ pub mod pallet {
...
+	pub fn DefaultFeeCurrency<T: Config>() -> T::AssetId {
+		T::NativeAssetId::get()
+	}
+
 	/// Account currency map
 	#[pallet::storage]
-	#[pallet::getter(fn get_currency)]
-	pub type AccountCurrencyMap<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, AssetIdOf<T>, OptionQuery>;
+	#[pallet::getter(fn account_currency)]
+	pub type AccountCurrency<T: Config> =
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:146]: this is major change that breaks API

File: pallets/transaction-multi-payment/src/lib.rs [hunk 4]
```diff
@@ -172,34 +131,29 @@ pub mod pallet {
...
 	/// Curated list of currencies which fees can be paid mapped to corresponding fallback price
 	#[pallet::storage]
 	#[pallet::getter(fn currencies)]
-	pub type AcceptedCurrencies<T: Config> = StorageMap<_, Twox64Concat, AssetIdOf<T>, Price, OptionQuery>;
-
-	/// Asset prices from the spot price provider or the fallback price if the price is not available. Updated at the beginning of every block.
-	#[pallet::storage]
-	#[pallet::getter(fn currency_price)]
-	pub type AcceptedCurrencyPrice<T: Config> = StorageMap<_, Twox64Concat, AssetIdOf<T>, Price, OptionQuery>;
+	pub type AcceptedCurrencies<T: Config> = StorageMap<_, Twox64Concat, AssetIdOf<T>, (), OptionQuery>;
```
Comment (mrq1911) [inline pallets/transaction-multi-payment/src/lib.rs:152]: this breaks API as well

File: runtime/hydradx/src/system.rs [hunk 5]
```diff
@@ -471,12 +473,11 @@ impl pallet_transaction_payment::Config for Runtime {
...
 impl pallet_transaction_multi_payment::Config for Runtime {
 	type RuntimeEvent = RuntimeEvent;
-	type AcceptedCurrencyOrigin = SuperMajorityTechCommittee;
-	type Currencies = Currencies;
-	type RouteProvider = Router;
-	type OraclePriceProvider = OraclePriceProvider<AssetId, EmaOracle, LRNA>;
+	type AssetId = AssetId;
+	type Balance = Balance;
+	type AuthorityOrigin = SuperMajorityTechCommittee;
+	type PriceProvider = OmnipoolSpotPriceProvider<Runtime>;
```
Comment (mrq1911) [inline runtime/hydradx/src/system.rs:479]: with this we support only assets that are in omnipool as payment assets, we should support every asset we have price for in oracle and aditionally even new assets that dont have any price in oracle yet (till shitcoin casino)

**Review by mrq1911 — CHANGES_REQUESTED** [review]

File: runtime/adapters/src/lib.rs [hunk 6]
```diff
@@ -153,7 +156,17 @@ impl<
 		);
 		let (asset_loc, price) = self.get_asset_and_price(&payment).ok_or(XcmError::AssetNotFound)?;
 		let fee = ConvertWeightToFee::weight_to_fee(&weight);
-		let converted_fee = price.checked_mul_int(fee).ok_or(XcmError::Overflow)?;
+		//let converted_fee = price.checked_mul_int(fee).ok_or(XcmError::Overflow)?;
```
Comment (apopiak) [inline runtime/adapters/src/lib.rs:159]: commented code

File: runtime/adapters/src/lib.rs [hunk 7]
```diff
@@ -178,7 +191,15 @@ impl<
...
 		if let Some(((asset_loc, price), amount)) = self.paid_assets.iter_mut().next() {
-			let converted_fee: u128 = price.saturating_mul_int(fee).saturated_into();
+			let (n, d) = (*price).into();
+			let converted_fee = multiply_by_rational_with_rounding(
+				fee.saturated_into(),
+				n,
+				d,
+				sp_arithmetic::per_things::Rounding::Up,
+			)?;
+			//let converted_fee = price.saturating_mul_int(fee).saturated_into();
```
Comment (apopiak) [inline runtime/adapters/src/lib.rs:201]: comment

File: runtime/adapters/src/lib.rs [hunk 8]
```diff
@@ -966,3 +979,21 @@ where
...
+where
+	SP: PriceProvider<AssetId, Price = FixedU128>,
+	P: pallet_transaction_multi_payment::Config,
+	<P as pallet_transaction_multi_payment::Config>::AssetId: From<u32>,
+{
+	fn price(currency: AssetId) -> Option<FixedU128> {
+		if currency == CORE_ASSET_ID {
+			Some(FixedU128::one())
+		} else {
+			//pallet_transaction_multi_payment::Pallet::<P>::get_currency_price(currency.into())
```
Comment (apopiak) [inline runtime/adapters/src/lib.rs:995]: comment

**Review by apopiak — COMMENTED** [review]
I guess you were planning to remove the comments later, so I stopped pointing them out at some point ^^

**Comment by enthusiastmartin** [comment]
closing this..due to recent changes in multi payment pallet - this refactor is no longer possible, or hard to re-refactor.

**Pull request closed without merging**